### PR TITLE
Merge AttackFrontal into AttackFollow

### LINKED ISSUE #10855: Support for turret firing arcs
Request. Turrets currently only function with 360 degree rotation. Firing arcs would increase flexibility with unit designs.

### PR DESCRIPTION
**Pull request by tovl**
Fixes #2327
Fixes #10855

This is the first step of the Attack* refactoring. The general aim is to decrease the number of distinct attack traits/activities and increase conformity in the remaining ones in order to increase maintainability as well as flexibility.

`AttackFollow` has been augmented to include turning to face the target if necessary and the option to disallow firing while moving. This means `AttackFollow` now completely supersedes `AttackFrontal`. This also opens up a couple of new possibilities:
- Turreted units with limited firing arcs
- Turreted units that have to stop to attack
- non-turreted units that can attack while moving.

The last two are demonstrated in the final two commits. These are meant primarily as testcases, but they might actually be interesting gameplay additions for the official mods.

Note that `AttackSwallow.SwallowTarget` inherited from the now defunct `Attack` activity. I've merged the two together omitting the obviously irrelevant parts as a temporary stopgap. I expect that this could be refactored further, but I want to defer that to a future PR.

### CODE REVIEW

**Comment by pchote** [comment]
Are there any objections to deferring this to Next + 1?

**Comment by matjaeck** [comment]
I think it's a good idea. Testing has become more difficult because the actual state of things has in many areas not stabilized yet, but already differs a lot from last release. Establishing a new status quo would help.

**Comment by reaperrr** [comment]
No objection, finding/fixing all of bleed's regressions will likely take several weeks as it is.

**Comment by tovl** [comment]
Rebased

**Comment by teinarss** [comment]
Needs a rebase.

File: OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs]: Consider change this to private set property

Comment (tovl) [reply OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs]: Refactored a bit so that making it public is no longer necessary.

**Comment by teinarss** [comment]
Found a regression: Clicking on enemy actor is not working anymore, the actor doesn't start moving towards the enemy actor.

**Comment by tovl** [comment]
Fixed.

**Comment by teinarss** [comment]
Crashing when trying to get eaten by a sandworm.

```
System.InvalidOperationException
  HResult=0x80131509
  Message=Sequence contains no elements
  Source=System.Core
  StackTrace:
   at System.Linq.Enumerable.Min[TSource](IEnumerable`1 source)
   at OpenRA.Mods.D2k.Traits.AttackSwallow.SwallowTarget..ctor(Actor self, Target target, Boolean allowMovement, Boolean forceAttack) in C:\projects\OpenRa1\OpenRA.Mods.D2k\Traits\AttackSwallow.cs:line 128
   at OpenRA.Mods.D2k.Traits.AttackSwallow.GetAttackActivity(Actor self, Target newTarget, Boolean allowMove, Boolean forceAttack, Nullable`1 targetLineColor) in C:\projects\OpenRa1\OpenRA.Mods.D2k\Traits\AttackSwallow.cs:line 84
   at OpenRA.Mods.Common.Traits.AttackBase.AttackTarget(Target target, Boolean queued, Boolean allowMove, Boolean forceAttack, Nullable`1 targetLineColor) in C:\projects\OpenRa1\OpenRA.Mods.Common\Traits\Attack\AttackBase.cs:line 402
   at OpenRA.Mods.D2k.Traits.Sandworm.RescanForTargets(Actor self) in C:\projects\OpenRa1\OpenRA.Mods.D2k\Traits\Sandworm.cs:line 92
   at OpenRA.Mods.D2k.Traits.Sandworm.OpenRA.Traits.ITick.Tick(Actor self) in C:\projects\OpenRa1\OpenRA.Mods.D2k\Traits\Sandworm.cs:line 78
   at OpenRA.World.<>c.<Tick>b__107_0(TraitPair`1 x) in C:\projects\OpenRa1\OpenRA.Game\World.cs:line 457
   at OpenRA.WorldUtils.DoTimed[T](IEnumerable`1 e, Action`1 a, String text) in C:\projects\OpenRa1\OpenRA.Game\WorldUtils.cs:line 78
   at OpenRA.World.Tick() in C:\projects\OpenRa1\OpenRA.Game\World.cs:line 457
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in C:\projects\OpenRa1\OpenRA.Game\Game.cs:line 610
   at OpenRA.Game.LogicTick() in C:\projects\OpenRa1\OpenRA.Game\Game.cs:line 637
   at OpenRA.Game.Loop() in C:\projects\OpenRa1\OpenRA.Game\Game.cs:line 794
   at OpenRA.Game.Run() in C:\projects\OpenRa1\OpenRA.Game\Game.cs:line 834
   at OpenRA.Game.InitializeAndRun(String[] args) in C:\projects\OpenRa1\OpenRA.Game\Game.cs:line 261
   at OpenRA.Program.Main(String[] args) in C:\projects\OpenRa1\OpenRA.Game\Support\Program.cs:line 31
```

**Comment by tovl** [comment]
Fixed.

**Review by teinarss — DISMISSED** [review]

**Review by abcdefg30 — COMMENTED** [review]
Doesn't this want an update rule?

**Comment by tovl** [comment]
Added update rule.

**Comment by tovl** [comment]
Rebased. Also, where does the "Has dependencies" label come from?

**Comment by pchote** [comment]
The dependency is on the stable release being pushed, to avoid diverging any further and making any remaining regression fixes more difficult.

**Comment by tovl** [comment]
Ok. Makes sense.

**Comment by reaperrr** [comment]
Release is tagged, I think we can remove the Has deps label now.

**Comment by tovl** [comment]
This has been sitting here unmerged for months with a +1; Can we please just get this over with now?

**Comment by pchote** [comment]
Does anything here need to change after #17413/#17418? I assume not, but just want to check before I start digging into this.

**Comment by tovl** [comment]
No, that will have to be dealt with in #16739

File: OpenRA.Mods.Common/Activities/AttackActivity.cs
Comment (reaperrr) [inline OpenRA.Mods.Common/Activities/AttackActivity.cs]: Any chance you could split the `Attack` -> `AttackActivity` rename/replacement to a separate commit (maybe rename the old AttackActivity to OldAttackActivity temporarily for the transition)?

My concern here is that the way the commit is now, comparing the diff between the old Attack.cs and this is tricky, and risks silent rebase conflicts with bleed.

Comment (tovl) [reply OpenRA.Mods.Common/Activities/AttackActivity.cs]: It's actually the other way around. `Attack` is not renamed to `AttackActivity`; It is the `AttackActivity` from `AttackFollow` that has been split off into it's own file with some small adjustments. the old `Attack` is completely defunct.

File: OpenRA.Mods.D2k/Traits/AttackSwallow.cs
Comment (pchote) [inline OpenRA.Mods.D2k/Traits/AttackSwallow.cs]: Likewise here - most of the old `AttackFrontal` logic - ranges and facings - isn't actually needed for the sandworms.

File: OpenRA.Mods.Cnc/Traits/Attack/AttackLeap.cs
Comment (reaperrr) [inline OpenRA.Mods.Cnc/Traits/Attack/AttackLeap.cs]: I don't think we use this approach (`new` key) anywhere else, but I don't remember if there was a technical reason for that, maybe @pchote can shed some light on this.

File: OpenRA.Mods.Common/UpdateRules/Rules/20190314/RemoveAttackFrontal.cs
Comment (pchote) [inline OpenRA.Mods.Common/UpdateRules/Rules/20190314/RemoveAttackFrontal.cs]: We have a lot more ex-`AttackFrontal` actors than `AttackOmni` etc.

IMO it would make the code a lot more understandable if we could change the default value in `AttackBase` to 0 and likewise pushed all the `TargetInFiringArc`s into the base `CanAttack` implementation. The few actors that really do want to disable firing arcs can then specify `FacingTolerance: 128`.

**Review by pchote — COMMENTED** [review]
I haven't looked at the merged activity in detail yet or tested this ingame - comments so far are based on a first read through the diff.

File: OpenRA.Mods.Common/Activities/AttackActivity.cs
Comment (tovl) [reply OpenRA.Mods.Common/Activities/AttackActivity.cs]: If it helps with reviewing, the changes compared to the old `AttackActivity` are:
- The addition of the block beneath "Turn to face the target if required"
- the addition of the `canMove` bool and associated changes.

Anything else could be a potential rebase conflict, although after half a year I am a bit fuzzy on the details.

File: OpenRA.Mods.Common/UpdateRules/Rules/20190314/RemoveAttackFrontal.cs
Comment (tovl) [reply OpenRA.Mods.Common/UpdateRules/Rules/20190314/RemoveAttackFrontal.cs]: Every turreted actor we have currently has `FacingTolerance: 128`

Comment (pchote) [reply OpenRA.Mods.Common/UpdateRules/Rules/20190314/RemoveAttackFrontal.cs]: I consider that to be a large part of the problem - `FacingTolerance` should be used consistently for all actors, not used as a firing cone for frontal units and a turret allowance angle for turreted units.

File: OpenRA.Mods.Cnc/Traits/Attack/AttackLeap.cs
Comment (tovl) [reply OpenRA.Mods.Cnc/Traits/Attack/AttackLeap.cs]: It is too long ago now, for me to remember what the reason for choosing this approach was at the time.

**Comment by reaperrr** [comment]
Both testcases work as expected, didn't notice any issues while watching an RA skirmish, either.
As far as I'm concerned, LGTM once @pchote is happy enough with the code changes.

File: OpenRA.Mods.Cnc/Traits/Attack/AttackLeap.cs
Comment (pchote) [reply OpenRA.Mods.Cnc/Traits/Attack/AttackLeap.cs]: We mainly don't use this approach for consistency with all the existing code that never used this approach. IMO we probably should start using it, and at some point fix the few existing cases that use variable name gymnastics to avoid it.

File: OpenRA.Mods.Common/UpdateRules/Rules/20190314/RemoveAttackFrontal.cs
Comment (pchote) [reply OpenRA.Mods.Common/UpdateRules/Rules/20190314/RemoveAttackFrontal.cs]: I suspect it would be easier to implement and review if we could fix this in a separate PR targeting bleed, then rebase this one on top.

Comment (tovl) [reply OpenRA.Mods.Common/UpdateRules/Rules/20190314/RemoveAttackFrontal.cs]: We currently don't have firing cones specifically for turrets; only for the bodies they are attached to. This is consistent for all actors. As it is we don't actually have proper turret allowance angles, and properly fixing that is something I consider something for a follow up. The aim of these first three PRs was to first apply some order to the jungle of attack traits we have before moving on to finetuning behaviours.

Comment (pchote) [reply OpenRA.Mods.Common/UpdateRules/Rules/20190314/RemoveAttackFrontal.cs]: > We currently don't have firing cones specifically for turrets; only for the bodies they are attached to.

Thanks for the correction, this is what I meant.

Trying to order the jungle up front then deal with the exceptions in followups didn't go very well during the activity overhaul, so i'd really like to avoid the same things repeating here.

Would we break any usecases by starting with a PR that changes turreted units to allow target acquisition from any angle and hooks up `FacingTolerance` as a firing cone?

Comment (tovl) [reply OpenRA.Mods.Common/UpdateRules/Rules/20190314/RemoveAttackFrontal.cs]: I don't know if it would break any usecases, but it would break my back.

I have zero motivation to change the whole plan after half a year of waiting. In fact, I'm starting to get burned out on this project as a whole, because of continued unreasonable demands like this. For what it's worth, I came very close to quitting after the whole move-undeploy debacle; I don't want to go through that bullshit again.

I really need to take a break from this project, but I would rather finish up any PRs that are still open in a tidy manner first. So please, accept this thing with minor corrections or close forever.

Comment (reaperrr) [reply OpenRA.Mods.Common/UpdateRules/Rules/20190314/RemoveAttackFrontal.cs]: I can't comment on individual change requests, but generally speaking, I can understand the getting burned out part. On average we're definitely taking far too long to give some PRs the attention they deserve, and that's pretty frustrating for everyone involved (fwiw, I'm in the same boat, I've had PRs sitting unreviewed for even longer in the past).

At least part of the problem for the lack of reviews in recent months, though, was that since release 20190314 had some show-stopperish issues (some of them outside our control, like Linux distros shipping broken Mono versions incompatible with OpenRA), we had to "hurry up" with getting another stable release out, which meant focusing on bugfixes over more refactorings.
Only that some of the larger refactorings we already merged turned that 'hurrying' into a ~5 month drag of almost nothing but bug-fixing, which - at least on my end, but I guess for everyone else involved, too - was absolute torture in terms of motivation.
tl;dr: The reason why I reduced my reviews to the bare minimum necessary to get the release out of the door over the last ~3-4 months was that I was, in fact, burned out, too. And I assume @pchote has been in a similar boat.

I'm afraid there isn't an easy near-term solution, other than taking a break when it gets too bad, since the core issue it boils down to is essentially lack of manpower.

We could consider (and have already discussed it in the past) shifting all official mods other than TS into maintenance mode (maybe keep them on bleed for testcases and potential future major updates, but otherwise freeze their official state in prep with only minor bugfix/maintenance/balance updates when sensible or necessary), so we could focus more on bleed in general and TS/next-gen mods in particular.

However, a) that increases the risk of accumulating too many subliminal bugs and other issues that only dedicated players would spot, making the following bugfix phase potentially even more devastating, and b) with the shroud perf improvement and @teinarss' HPA on the horizon, I feel that at least the current prep branch would be too early to go down that route. /semi-off-topic-wall-o-text

Comment (matjaeck) [reply OpenRA.Mods.Common/UpdateRules/Rules/20190314/RemoveAttackFrontal.cs]: Not sure if the circumstances are right to start this attack refactor. Reviewing has not even really started (PR was on hold for the last months) and the level of frustration is already alarming. Better drop the patches for now and let everyone take a break to recover if that is needed.

Comment (pchote) [reply OpenRA.Mods.Common/UpdateRules/Rules/20190314/RemoveAttackFrontal.cs]: @tovl: To be honest, your response hurts. The last release was a kick in the teeth for all of us, and I too am feeling burned out and still rather resentful about organizing most of my scarce free time over the summer around OpenRA. I invested more time than I was happy to give, at the expense of my personal and professional life, because it felt like I was the only person who really cared about the impact that the bugs in the previous release were having on our player community - and "just one more weekend" should have been enough for the stable release that never quite happened.

The situation sucks, but we all have to deal with it the best we can. I still feel that delaying this PR was the right choice and i'm sure that #17413 would have been a lot harder to resolve if this, #16739, and five months worth of followup work had been merged to bleed before somebody found it.

The overall direction of the attack refactor looks good, but its not fair or reasonable to try and tackle this in the same way as the activity rework. That was an awesome technical achievement, but it destroyed at least mine and @reaperrr's motivation, and I am still struggling to find any motivation to get back into reviewing. A large part of the problem was that the initial PRs were too ambitious, and broke too many unexpected assumptions that took too much work to identify after the fact and fix. I remember you agreeing on this point a few months ago.

The way to avoid this is, having worked out the broad shape of the refactoring, to take the extra time and extra PRs to rework the surrounding code and fix assumptions ahead of time. This focuses review and testing efforts, and makes it a lot easier to spot and fix any unexpected regressions before they become monstrous. Look at any of the big features I've implemented over the last few years to see that I hold myself to this standard as much (probably more) than anyone else.

Given this, I feel that my request about the `FacingTolerance` is completely reasonable. I would bet that the behaviour differences there would otherwise become a problem before the end of this, and that it will be a nightmare to fix if we wait until all of our combat units share 75% of their logic before doing so.

If you still think I'm being unreasonable then I agree that its probably for the best that we close this for now (but definitely park it on the Future milestone) until our collective spirits have raised enough to tackle it again.

**Comment by GraionDilach** [comment]
Another important milestone sacrificed in the name of the shortsightedness.

**Pull request closed without merging**
